### DISCUSSION TITLE
Why is `lm_head` in ESM-2 (35M) (480, 480)?

### DISCUSSION THREAD

**Discussion by jasperhyp** [facebookresearch/esm#416] [General] [answered]
Hi!  Please correct me if I am wrong, but I saw [here](https://github.com/facebookresearch/esm/blob/4e0ebb7a7b875ef40178cbb11e830eb5859b4180/esm/model/esm2.py#L71) that `lm_head` is (480, 33).  However, when loading the model using `model, alphabet = esm.pretrained.esm2_t12_35M_UR50D()`, `model.lm_head` is actually having a dense layer of (480, 480).  Why the discrepancy?

**Comment by tomsercu** [comment, marked as answer]
When you look at the [definition of lm_head](https://github.com/facebookresearch/esm/blob/4e0ebb7a7b875ef40178cbb11e830eb5859b4180/esm/modules.py#L298) you'll see the forward pass is dense -> nonlinearities -> output layer. So the 480,480 is the first of a 2layer MLP.
And the output layer is shared with the transformer input token embedding.

**Reply by jasperhyp** [reply]
Thank you!
